sound/soc/sof/topology.c

```diff
@@ -2148,10 +2065,22 @@ static int sof_complete(struct snd_soc_component *scomp)
 	struct snd_sof_widget *swidget, *comp_swidget;
 	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
 	const struct ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	struct snd_sof_control *scontrol;
 	int ret;
 
+	/* first update all control IPC structures based on the IPC version */
+	if (ipc_tplg_ops->control_setup)
+		list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
+			ret = ipc_tplg_ops->control_setup(sdev, scontrol);
+			if (ret < 0) {
+				dev_err(sdev->dev, "failed updating IPC struct for control %s\n",
+					scontrol->name);
+				return ret;
+			}
+		}
```

Comment:
In the previous version this was done as the last thing, after `sof_set_up_pipelines()`
The dynamic_pipeline_widget flag for example is applied later, which was checked previously, if I recall right...

Reply:
@ujfalusi that was a different PR for a completely different thing to read back control data. This one is about setting up the control IPC control message. Has to be done first

Reply:
Too many flying around ;)